volume/drivers/coprhd/coprhd.go

```diff
@@ -186,71 +274,205 @@ func (d *driver) Alerts(volumeID string) (*api.Alerts, error) {
 }
 
 func (d *driver) Attach(volumeID string) (path string, err error) {
-	return "", nil
+	volume, err := d.GetVol(volumeID)
+	if err != nil {
+		return "", fmt.Errorf("Volume %s could not be located", volumeID)
+	}
+
+	coprVolume, err := d.client.Volume().
+		WWN(volumeID).
+		Query()
+	if err != nil {
+		return "", err
+	}
+
+	group, err := d.client.Export().
+		Name(coprVolume.Name).
+		Initiators(d.initiator.Id).
+		Volumes(coprVolume.Id).
+		Project(coprVolume.Project.Id).
+		Array(coprVolume.VArray.Id).
+		Create()
+	if err != nil {
+		return "", err
+	}
+
+	protocol := d.initiator.Protocol
+	device := ""
+
+	switch protocol {
+	case coprhd.InitiatorTypeScaleIO:
+		device, err = d.waitAttachDevice(volumeID, time.Second*180)
+		if err != nil {
+			return "", err
+		}
+	default:
+		return "", fmt.Errorf("Unsupported initiator protocol")
+	}
+
+	volume.AttachedOn = group.Id
+	volume.DevicePath = device
+	volume.Status = api.VolumeStatus_VOLUME_STATUS_UP
+
+	if err := d.UpdateVol(volume); err != nil {
+		return "", err
+	}
+	return device, nil
 }
 
 func (d *driver) Detach(volumeID string) error {
+	v, err := d.GetVol(volumeID)
+	if err != nil {
+		return fmt.Errorf("Volume is not attached")
+	}
+
+	export := v.AttachedOn
+	if export != "" {
+		err = d.client.Export().
+			Id(export).
+			Delete()
+		if err != nil {
+			return err
+		}
+	}
+
+	v.AttachedOn = ""
```

Comment:
this may be a bigger openstorage question, but what guarantees are there that this will be executed properly concurrently?